drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -625,26 +651,66 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
 	return 0;
 }
 
-int usb_dc_ep_read_continue(uint8_t ep)
+int usb_dc_ep_read_continue(const uint8_t ep)
 {
-	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
-	size_t bytes_received;
+	const struct udc_rpi_ep_state *const ep_state = udc_rpi_get_ep_state(ep);
+	bool arm_out_endpoint = false;
 
 	if (!ep_state || !USB_EP_DIR_IS_OUT(ep)) {
 		LOG_ERR("Not valid endpoint: %02x", ep);
 		return -EINVAL;
 	}
+	if (ep == USB_CONTROL_EP_OUT && state.setup_available) {
+		const struct usb_setup_packet *setup = (const void *)&usb_dpram->setup_packet;
+
+		LOG_DBG("EP0 setup (wLength=%d, is_to_device=%d)",
+			setup->wLength, usb_reqtype_is_to_device(setup));
+		if (setup->wLength != 0U) {
+			/* In the case of a control transfer, we want to prime the OUT endpoint
+			 * exactly once, to either:
+			 * 1) in the to_device case, to receive the data (only if wLength is not 0)
+			 * 2) in the to_host case, to receive a 0-length status stage transfer
+			 *    (only valid if wLength is not 0)
+			 * Note that when wLength = 0, the status stage transfer is always an IN
+			 * type so we don't need to consider that case.
+			 */
+			arm_out_endpoint = true;
+			state.control_transfer_bytes_transferred = 0;
+		}
+		state.setup_available = false;
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(ep);
+
+		LOG_DBG("EP %d received %d offset: %d", ep, bytes_received, ep_state->read_offset);
+		if (ep == USB_CONTROL_EP_OUT) {
+			const struct usb_setup_packet *setup =
+				(const void *)&usb_dpram->setup_packet;
+			if (usb_reqtype_is_to_device(setup)) {
+				if (state.control_transfer_bytes_transferred + ep_state->read_offset
+					< setup->wLength) {
+					/* If no more data in the buffer, but we're still waiting
+					 * for more, start a new read transaction.
+					 */
+					if (bytes_received == ep_state->read_offset) {
```

Comment:
I don't understand this line. When can it be not equal? Is this just checking for a short data packet (i.e. not bMaxPacketSize0)?

Reply:
This allows us to receive for example a single packet of 64 bytes, but then proceed to read it out in multiple pieces by calling usb_dc_ep_read() repeatedly with smaller max_data_len. When there is no data remaining in the 64 byte buffer, only then can we re-arm the endpoint, otherwise a new packet coming in could overwrite it. The only exception to this behavior is when reading in the 8 bytes of data associated with a setup transaction of a control transfer, in which case any bytes left over will be lost, all 8 bytes must be read at once if you need them. Note that this behavior is the same as it was before my PR, I did not change it. You will find the same comparison here: https://github.com/zephyrproject-rtos/zephyr/pull/54030/files#diff-927fe840a66cec50c33e92952dbef313db723ed7a43bdea4603da8d3c312373dL646

Reply:
This is strange and somewhat questionable. Transfers less than max packet size generally end transfer and allowing application to read less than max packet size of received data is somewhat weird. However, because it was there already I'm fine with it remaining there for the time being (especially that this PR does fix real issues).